test/read.jl

```diff
@@ -0,0 +1,314 @@
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, Base.Filesystem.File)
+close(s)
+push!(l, ("File", io))
+
+
+# IOStream
+io = (text) -> begin
+    write(filename, text)
+    open(filename)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOStream)
+close(s)
+push!(l, ("IOStream", io))
+
+
+# IOBuffer
+io = (text)->IOBuffer(text)
+s = io(text)
+@test isa(s, IO)
+@test isa(s, IOBuffer)
+close(s)
+push!(l, ("IOBuffer", io))
+
+
+function run_test_server(srv, text)
+    push!(tasks, @async begin
+        try
+            sock = accept(srv)
+            try
+                write(sock,text)
+            catch e
+                if typeof(e) != Base.UVError
+                    rethrow(e)
+                end
+            finally
+                close(sock)
+            end
+        finally
+            close(srv)
+        end
+    end)
+    yield()
+end
+
+
+# TCPSocket
+io = (text) -> begin
+    port, srv = listenany(rand(2000:4000))
+    run_test_server(srv, text)
+    connect(port)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, TCPSocket)
+close(s)
+push!(l, ("TCPSocket", io))
+
+
+# PipeEndpoint
+io = (text) -> begin
+    a = "\\\\.\\pipe\\uv-test-$(randstring(6))"
+    b = joinpath(dir, "socket-$(randstring(6))")
+    socketname = @windows ? a : b
+    srv = listen(socketname)
+    run_test_server(srv, text)
+    connect(socketname)
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, Base.PipeEndpoint)
+close(s)
+push!(l, ("PipeEndpoint", io))
+
+
+#FIXME See https://github.com/JuliaLang/julia/issues/14747
+#      Reading from open(::Command) seems to deadlock on Linux/Travis
+#=
+@windows ? nothing : begin
+
+# Windows type command not working?
+# See "could not spawn `type 'C:\Users\appveyor\AppData\Local\Temp\1\jul3516.tmp\file.txt'`"
+#https://ci.appveyor.com/project/StefanKarpinski/julia/build/1.0.12733/job/hpwjs4hmf03vs5ag#L1244
+
+# Pipe
+io = (text) -> begin
+    write(filename, text)
+    open(`$(@windows ? "type" : "cat") $filename`)[1]
+#    Was open(`echo -n $text`)[1]
+#    See https://github.com/JuliaLang/julia/issues/14747
+end
+s = io(text)
+@test isa(s, IO)
+@test isa(s, Pipe)
+close(s)
+push!(l, ("Pipe", io))
+
+end
+=#
+
+
+open_streams = []
+function cleanup()
+    for s in open_streams
+        try close(s) end
+    end
+    empty!(open_streams)
+    for tsk in tasks
+        wait(tsk)
+    end
+    empty!(tasks)
+end
+
+verbose = false
+
+                                                                       cleanup()
+
+for (name, f) in l
+
+    io = ()->(s=f(text); push!(open_streams, s); s)
+
+    write(filename, text)
+
+    verbose && println("$name read...")
+    @test read(io(), UInt8) == read(IOBuffer(text), UInt8)
+    @test read(io(), UInt8) == read(filename, UInt8)
+    @test read(io(), Int) == read(IOBuffer(text), Int)
+    @test read(io(), Int) == read(filename,Int)
+                                                                       cleanup()
```

Comment:
I've just taken out most of the temporary `cleanup()` calls.
That was an attempt to figure out why CI was hanging (i.e. maybe because of too many tasks/file handles/sockets ...).